Fix GCS regex pattern [BW-494]

aednichols opened the pull request:
In [CROM-6338](https://broadworkbench.atlassian.net/browse/CROM-6338) Denis reports that Cromwell is unexpectedly failing to retry 503s and provides the following sample error:
```
"failures": [
  {
    "causedBy": [
      {
        "causedBy": [
          {
            "causedBy": [],
            "message": "503 Service Unavailable\nBackend Error"
          }
        ],
        "message": "Could not read from gs://broad-epi-cromwell/workflows/ChipSeq/ce6a5671-baf6-4734-a32b-abf3d9138e9b/call-epitope_classifier/memory_retry_rc: 503 Service Unavailable\nBackend Error"
      }
    ],
    "message": "[Attempted 1 time(s)] - IOException: Could not read from gs://broad-epi-cromwell/workflows/ChipSeq/ce6a5671-baf6-4734-a32b-abf3d9138e9b/call-epitope_classifier/memory_retry_rc: 503 Service Unavailable\nBackend Error"
  }
]
```

In https://github.com/broadinstitute/cromwell/issues/6154 @freeseek reports that Cromwell is unexpectedly failing to retry 504s and provides the following sample error:
```
{
  "causedBy": [
    {
      "causedBy": [
        {
          "message": "504 Gateway Timeout\nGET https://storage.googleapis.com/download/storage/v1/b/mccarroll-mocha/o/cromwell%2Fcromwell-executions%2Fmocha%2F86d47e9a-5745-4ec0-b4eb-0164f073e5f4%2Fcall-idat2gtc%2Fshard-73%2Frc?alt=media",
          "causedBy": []
        }
      ],
      "message": "Could not read from gs://mccarroll-mocha/cromwell/cromwell-executions/mocha/86d47e9a-5745-4ec0-b4eb-0164f073e5f4/call-idat2gtc/shard-73/rc: 504 Gateway Timeout\nGET https://storage.googleapis.com/download/storage/v1/b/mccarroll-mocha/o/cromwell%2Fcromwell-executions%2Fmocha%2F86d47e9a-5745-4ec0-b4eb-0164f073e5f4%2Fcall-idat2gtc%2Fshard-73%2Frc?alt=media"
    }
  ],
  "message": "[Attempted 1 time(s)] - IOException: Could not read from gs://mccarroll-mocha/cromwell/cromwell-executions/mocha/86d47e9a-5745-4ec0-b4eb-0164f073e5f4/call-idat2gtc/shard-73/rc: 504 Gateway Timeout\nGET https://storage.googleapis.com/download/storage/v1/b/mccarroll-mocha/o/cromwell%2Fcromwell-executions%2Fmocha%2F86d47e9a-5745-4ec0-b4eb-0164f073e5f4%2Fcall-idat2gtc%2Fshard-73%2Frc?alt=media"
}
```
Our regexes did not allow for the `\n` in the Google errors. I believe this bug came about when copy-pasting to create the test cases.

aednichols commented:
I think this PR may be fake news.

The real problem may be that [the `IO` that runs these file operations](https://github.com/broadinstitute/cromwell/blob/develop/engine/src/main/scala/cromwell/engine/io/nio/NioFlow.scala#L42-L50) never calls `isRetryable`.

cjllanwarne commented on engine/src/main/scala/cromwell/engine/io/RetryableRequestSupport.scala:
```diff
@@ -59,17 +59,17 @@ object RetryableRequestSupport {
 
   def isGcs500(failure: Throwable): Boolean = {
-    val serverErrorPattern = ".*Could not read from gs.+500 Internal Server Error.*"
+    val serverErrorPattern = ".*500 Internal Server Error.*"
```
TOL would a `String.contains` be a simpler/better way to express this intention?

cjllanwarne commented on common/src/main/scala/common/util/IORetry.scala:
```diff
@@ -33,23 +33,23 @@ object IORetry {
                       backoff: Backoff,
-                      isTransient: Throwable => Boolean = throwableToFalse,
-                      isFatal: Throwable => Boolean = throwableToFalse,
+                      isRetryable: Throwable => Boolean = throwableToTrue,
```
TOL it's a little confusing that we have two functions to return almost the same information here. And because they're separate, it's possible that they combine to produce an invalid state. Maybe a `Retryability` category case class with three options could more precisely capture the designation here?

```scala
sealed trait IoErrorRetryabilityCategory
case object FatalIoError extends IoRetryabilityCategory
case object RetryableIoError extends IoRetryabilityCategory
case object InfinitelyRetryableIoError extends IoRetryabilityCategory
```

Then you could have a single input function like:
```scala
retryabilityCategorizationFunction: Throwable => IoErrorRetryabilityCategory
```

cjllanwarne commented on engine/src/main/scala/cromwell/engine/io/RetryableRequestSupport.scala:
```diff
@@ -51,25 +51,30 @@ object RetryableRequestSupport {
-    case _ => false
+  def isInfinitelyRetryable(failure: Throwable): Boolean = {
+
+    // Retry forever because eventually the user will have quota capacity again
```
I have a feeling that "out of quota" is sometime used to indicate "you don't have any of that resource in this region". I guess it's ok to keep trying infinitely since we already were... but I'm nevertheless a little nervous about this!

aednichols replied:
According to the docs [0], `429` responses from GCS should be met with "truncated exponential backoff" [1].

I wasn't sure what the "truncated" part meant, and Wikipedia helpfully explains [2]:

> The 'truncated' simply means that after a certain number of increases, the exponentiation stops; i.e. the retransmission timeout reaches a ceiling, and thereafter does not increase any further.

They don't mention an absolute limit on time to retry or number of attempts, which implicitly leads me to believe that infinite is OK.

[0] https://cloud.google.com/storage/docs/json_api/v1/status-codes#429_Too_Many_Requests
[1] https://cloud.google.com/storage/docs/exponential-backoff
[2] https://en.wikipedia.org/wiki/Exponential_backoff

cjllanwarne reviewed (approved):
I like the renaming and this all seems logically correct, so you can keep the 👍  from before, but you don't seem to have addressed your epiphany (as far as I can tell?)

> The real problem may be that the IO that runs these file operations never calls isRetryable.

aednichols reviewed (commented):
> The real problem may be that the IO that runs these file operations never calls isRetryable.

> you don't seem to have addressed your epiphany

Apologies, it was a bit unfair to request re-review without clarifying. It was a three-day weekend 😄 

The outcome is that I did not realize that `isFatal` in `IORetry.scala` was the inverse of `isRetryable`.

aednichols commented:
This PR still has a problem though, which is that the `isGcs50X`  functions now make no attempt to check whether the code is from GCS or not. Previously, we checked the message for a string component emitted by `EvenBetterPathMethods` (which narrows it down to file operations) as well as the `gs` prefix.

aednichols commented:
It's suspicious to me that in the Terra logs we have many, many instances of
```
"[Attempted 1 time(s)] - IOException: Could not read from gs://"
```
in the last 30 days and zero instances of
```
"[Attempted 2 time(s)] - IOException: Could not read from gs://"
```
---
There are also workflows that fail while referencing `Attempted 1 time` for an error that should be retried like `504`:
```
"WorkflowManagerActor Workflow 02279d18-f3a9-43d0-b859-7f980f7e433e failed (during ExecutingWorkflowState): cromwell.engine.io.IoAttempts$EnhancedCromwellIoException: [Attempted 1 time(s)] - IOException: Could not read from gs://fc-secure-2d4b04e6-f83f-481b-a967-b211aad522c1/6125ce41-0102-4491-b739-8a347317f7c7/starworkflow/02279d18-f3a9-43d0-b859-7f980f7e433e/call-deletionCallSTAR/rc: 504 Gateway Timeout
GET https://storage.googleapis.com/download/storage/v1/b/fc-secure-2d4b04e6-f83f-481b-a967-b211aad522c1/o/6125ce41-0102-4491-b739-8a347317f7c7%2Fstarworkflow%2F02279d18-f3a9-43d0-b859-7f980f7e433e%2Fcall-deletionCallSTAR%2Frc?alt=media
Caused by: java.io.IOException: Could not read from gs://fc-secure-2d4b04e6-f83f-481b-a967-b211aad522c1/6125ce41-0102-4491-b739-8a347317f7c7/starworkflow/02279d18-f3a9-43d0-b859-7f980f7e433e/call-deletionCallSTAR/rc: 504 Gateway Timeout
GET https://storage.googleapis.com/download/storage/v1/b/fc-secure-2d4b04e6-f83f-481b-a967-b211aad522c1/o/6125ce41-0102-4491-b739-8a347317f7c7%2Fstarworkflow%2F02279d18-f3a9-43d0-b859-7f980f7e433e%2Fcall-deletionCallSTAR%2Frc?alt=media
	at cromwell.core.path.EvenBetterPathMethods$$anonfun$fileIoErrorPf$1.applyOrElse(EvenBetterPathMethods.scala:117)
	at cromwell.core.path.EvenBetterPathMethods$$anonfun$fileIoErrorPf$1.applyOrElse(EvenBetterPathMethods.scala:116)
	at map @ cromwell.engine.io.nio.NioFlow.handleSingleCommand(NioFlow.scala:69)
	at map @ cromwell.engine.io.nio.NioFlow.$anonfun$processCommand$5(NioFlow.scala:55)
	at flatMap @ cromwell.engine.io.nio.NioFlow.$anonfun$processCommand$1(NioFlow.scala:54)
Caused by: com.google.api.client.googleapis.json.GoogleJsonResponseException: 504 Gateway Timeout
GET https://storage.googleapis.com/download/storage/v1/b/fc-secure-2d4b04e6-f83f-481b-a967-b211aad522c1/o/6125ce41-0102-4491-b739-8a347317f7c7%2Fstarworkflow%2F02279d18-f3a9-43d0-b859-7f980f7e433e%2Fcall-deletionCallSTAR%2Frc?alt=media
	at com.google.api.client.googleapis.json.GoogleJsonResponseException.from(GoogleJsonResponseException.java:149)
	at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:112)
	at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:39)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest$1.interceptResponse(AbstractGoogleClientRequest.java:443)
	at com.google.api.client.http.HttpRequest.execute(HttpRequest.java:1108)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:541)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:474)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeMedia(AbstractGoogleClientRequest.java:502)
	at com.google.api.services.storage.Storage$Objects$Get.executeMedia(Storage.java:7029)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeMediaAsInputStream(AbstractGoogleClientRequest.java:645)
	at com.google.api.services.storage.Storage$Objects$Get.executeMediaAsInputStream(Storage.java:7024)
	at cromwell.filesystems.gcs.GcsPath.request$2(GcsPathBuilder.scala:213)
	at cromwell.filesystems.gcs.GcsPath.$anonfun$mediaInputStream$2(GcsPathBuilder.scala:218)
	at cromwell.filesystems.gcs.GcsPath.$anonfun$mediaInputStream$2$adapted(GcsPathBuilder.scala:218)
	at cromwell.filesystems.gcs.GcsEnhancedRequest$.$anonfun$recoverFromProjectNotProvided$1(GcsEnhancedRequest.scala:15)
	at map @ cromwell.filesystems.gcs.GcsPath.$anonfun$runOnEc$1(GcsPathBuilder.scala:233)
	at flatMap @ cromwell.filesystems.gcs.GcsPath.runOnEc(GcsPathBuilder.scala:232)
"
```

aednichols commented:
@cjllanwarne @salonishah11 this is really fixed now and ready for re-review. The Google errors provided by customers have `\n` in them which our regexes did not match.

Fixed the regex and updated test cases to match the actual error.

salonishah11 reviewed (approved):
💯
